File: src/services/translation/completeness/triggerDetector.ts

import { isInputComplete } from './inputDetector';

// 状态变量
let lastInputCheckTime = 0;
let lastCompleteText = '';
let consecutiveCompleteCount = 0;
let pauseTimer: NodeJS.Timeout | null = null;
let lastInputText = '';
let pauseCounter = 0;

/**
 * 检查文本是否符合翻译条件
 * 结合了文本长度、完整性和时间阈值的逻辑
 */
export function shouldTranslate(
  sourceText: string, 
  sourceLanguageCode: string,
  lastTranslatedText: string,
  isFirstTranslation: boolean
): boolean {
  const currentTime = Date.now();
  
  // 空文本不翻译
  if (!sourceText.trim()) {
    lastInputCheckTime = currentTime;
    consecutiveCompleteCount = 0;
    lastInputText = '';
    if (pauseTimer) {
      clearTimeout(pauseTimer);
      pauseTimer = null;
    }
    pauseCounter = 0;
    return false;
  }
  
  // 避免重复翻译相同的文本
  if (sourceText === lastTranslatedText && !isFirstTranslation) {
    return false;
  }
  
  // 检测用户是否停止输入（文本没有变化）
  const textUnchanged = sourceText === lastInputText;
  if (textUnchanged) {
    pauseCounter++;
  } else {
    pauseCounter = 0;
    lastInputText = sourceText;
    // 重置最后一次检查时间，用于计时器
    lastInputCheckTime = currentTime;
  }
  
  // 如果用户停止输入超过3次检查，认为可能已经完成输入
  const userPausedTyping = pauseCounter >= 3;
  
  // 检查用户是否暂停输入的时间
  const userPauseTime = currentTime - lastInputCheckTime;
  const wordCount = sourceText.trim().split(/\s+/).length;
  
  // 单词长度适中或较长时，应该更快触发翻译
  const isMediumWord = sourceText.trim().length >= 3;
  
  // 根据输入的单词数量和长度调整暂停阈值：
  // - 单个较短单词：800毫秒
  // - 单个中等长度以上单词：500毫秒
  // - 多个单词：需要更长时间确认是否完成，800毫秒
  const pauseThreshold = wordCount <= 1 ? (isMediumWord ? 500 : 800) : 800;
  const userPausedLongEnough = userPauseTime >= pauseThreshold;
  
  // 单个单词且已暂停足够时间就触发翻译
  if (wordCount <= 2 && userPausedLongEnough) {
    console.log("单词输入暂停触发翻译:", { 
      暂停时长: userPauseTime,
      单词: sourceText,
      暂停阈值: pauseThreshold
    });
    return true;
  }
  
  // 检查输入是否完整
  const isComplete = isInputComplete(sourceText, sourceLanguageCode);
  
  // 如果文本判断为完整
  if (isComplete) {
    // 如果文本内容与上次相同，且已被判定为完整，增加连续完整计数
    if (sourceText === lastCompleteText) {
      consecutiveCompleteCount += 1;
    } else {
      // 新的完整文本，重置计数
      consecutiveCompleteCount = 1;
      lastCompleteText = sourceText;
    }
    
    // 检查是否已经过了时间阈值或连续判定多次为完整
    // 时间阈值：如果是中文，设置更短的延迟，因为中文通常不需要太长的停顿判断
    const timeThreshold = sourceLanguageCode === 'zh' ? 500 : 600; // 降低阈值，提高响应速度
    const hasExceededTimeThreshold = (currentTime - lastInputCheckTime) >= timeThreshold;
    const shouldTriggeredByConsecutiveChecks = consecutiveCompleteCount >= 1; // 只要判定为完整就触发
    
    // 满足任一条件触发翻译
    if (hasExceededTimeThreshold || shouldTriggeredByConsecutiveChecks || userPausedTyping) {
      console.log("触发翻译：", { 
        时间阈值: hasExceededTimeThreshold, 
        连续完整: shouldTriggeredByConsecutiveChecks,
        用户停顿: userPausedTyping, 
        文本: sourceText 
      });
      
      // 重置状态
      lastInputCheckTime = currentTime;
      consecutiveCompleteCount = 0;
      pauseCounter = 0;
      return true;
    }
    
    // 更新时间戳以便下次检查
    lastInputCheckTime = currentTime;
  }
  
  // 如果文本不完整但长度大于2个字符，检查是否停顿了足够长的时间
  if (!isComplete && sourceText.trim().length >= 2 && userPausedLongEnough) {
    console.log("不完整文本停顿触发翻译:", { 
      停顿时长: userPauseTime, 
      文本: sourceText,
      暂停阈值: pauseThreshold 
    });
    pauseCounter = 0;
    return true;
  }
  
  // 不触发翻译
  return false;
}

File: backend/app/services/completeness/trigger_detector.py
# @AI-Generated
"""
触发检测器
"""
import time
import re
from .input_detector import is_input_complete

# 触发检测相关常量
PAUSE_THRESHOLD_SHORT = 600  # ms，短文本停顿阈值
PAUSE_THRESHOLD_LONG = 800   # ms，长文本停顿阈值
TIME_THRESHOLD_ZH = 500      # ms，中文时间阈值
TIME_THRESHOLD_OTHER = 600   # ms，其他语言时间阈值
CONSECUTIVE_COMPLETE_LIMIT = 1  # 连续完整检测次数
PAUSE_COUNTER_LIMIT = 3      # 连续未变动计数

class InputTriggerDetector:
    """
    输入触发检测器，封装状态，支持多用户/多会话
    """
    def __init__(self):
        self._last_input_check_time = 0
        self._last_complete_text = ''
        self._consecutive_complete_count = 0
        self._pause_counter = 0
        self._last_input_text = ''

    def reset_state(self):
        self._last_input_check_time = 0
        self._consecutive_complete_count = 0
        self._last_input_text = ''
        self._pause_counter = 0
        self._last_complete_text = ''

    async def should_translate(self, source_text: str, source_language_code: str, last_translated_text: str, is_first_translation: bool, llm_api_key: str = None) -> bool:
        current_time = int(time.time() * 1000)
        if not source_text.strip():
            self.reset_state()
            return False
        if source_text == last_translated_text and not is_first_translation:
            return False
        text_unchanged = source_text == self._last_input_text
        if not text_unchanged:
            self._last_input_text = source_text
            self._last_input_check_time = current_time
            self._pause_counter = 0
        else:
            self._pause_counter += 1
        user_paused_typing = self._pause_counter >= PAUSE_COUNTER_LIMIT
        user_pause_time = current_time - self._last_input_check_time
        word_count = len(source_text.strip().split())
        pause_threshold = PAUSE_THRESHOLD_SHORT if word_count <= 2 else PAUSE_THRESHOLD_LONG
        user_paused_long_enough = user_pause_time >= pause_threshold
        if user_paused_long_enough:
            return True
        is_complete = await is_input_complete(source_text, source_language_code, llm_api_key)
        if is_complete:
            if source_text == self._last_complete_text:
                self._consecutive_complete_count += 1
            else:
                self._consecutive_complete_count = 1
                self._last_complete_text = source_text
            time_threshold = TIME_THRESHOLD_ZH if source_language_code == 'zh' else TIME_THRESHOLD_OTHER
            has_exceeded_time_threshold = (current_time - self._last_input_check_time) >= time_threshold
            should_triggered_by_consecutive_checks = self._consecutive_complete_count >= CONSECUTIVE_COMPLETE_LIMIT
            if has_exceeded_time_threshold or should_triggered_by_consecutive_checks or user_paused_typing:
                self._last_input_check_time = current_time
                self._consecutive_complete_count = 0
                self._pause_counter = 0
                return True
            self._last_input_check_time = current_time
        if not is_complete and len(source_text.strip()) >= 2 and user_paused_long_enough:
            self._pause_counter = 0
            return True
        return False

    async def should_translate_ex(self, source_text: str, source_language_code: str, llm_api_key: str = None) -> dict:
        """
        增强版停顿/完整性检测
        :return: { should: bool, is_complete: bool }
        """
        if not hasattr(self, '_last_input_time'):
            self._last_input_time = int(time.time() * 1000)
        is_complete = await is_input_complete(source_text, source_language_code, llm_api_key)
        has_latin = bool(re.search(r'[a-zA-Z]', source_text))
        has_cjk = bool(re.search(r'[\u4e00-\u9fa5]', source_text))
        if has_latin and has_cjk:
            is_complete = False
        now = int(time.time() * 1000)
        last = self._last_input_time
        pause = now - last
        word_count = len(source_text.strip().split())
        is_single_word = word_count == 1
        if is_single_word and pause >= PAUSE_THRESHOLD_SHORT:
            self._last_input_time = now
            return {"should": True, "is_complete": True}
        if not is_complete and pause >= PAUSE_THRESHOLD_LONG:
            self._last_input_time = now
            return {"should": True, "is_complete": False}
        if is_complete:
            self._last_input_time = now
            return {"should": True, "is_complete": True}
        return {"should": False, "is_complete": is_complete}

# 单例兼容原有用法
_detector_instance = InputTriggerDetector()

async def should_translate(source_text: str, source_language_code: str, last_translated_text: str, is_first_translation: bool, llm_api_key: str = None) -> bool:
    """
    兼容原有API，使用单例
    """
    return await _detector_instance.should_translate(source_text, source_language_code, last_translated_text, is_first_translation, llm_api_key)

async def should_translate_ex(source_text: str, source_language_code: str, llm_api_key: str = None) -> dict:
    """
    兼容原有API，使用单例
    """
    return await _detector_instance.should_translate_ex(source_text, source_language_code, llm_api_key) 